README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# phantSEM

<!-- badges: start -->

[![R-CMD-check](https://github.com/argeorgeson/phantSEM/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/argeorgeson/phantSEM/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

The goal of phantSEM is to make it easier to create phantom variables,
which are variables that were not observed, for the purpose of
sensitivity analyses for structural equation models. The package allows
a user to test different combinations of covariances between the phantom
variable(s) and observed variables.

## Installation

You can install phantSEM from CRAN:

``` r
install.packages("phantSEM")
```

You can install the development version of phantSEM from
[GitHub](https://github.com/argeorgeson/phantSEM) with:

``` r
# install.packages("devtools")
devtools::install_github("argeorgeson/phantSEM")
```

## Example

This is an example that shows you how to use the package. Assume that
you have a simple mediation model with three variables, X, M2, and Y2.
You can use the phantSEM package to create phantom variables for
baseline observations of M and Y.

``` r
library(phantSEM)
## basic example code
covmatrix <- matrix(c(
  0.25, 0.95, 0.43,
  0.95, 8.87, 2.66,
  0.43, 2.66, 10.86
), nrow = 3, byrow = TRUE)
colnames(covmatrix) <- c("X", "M2", "Y2")

# lavann syntax for observed model
observed <- " M2 ~ X
              Y2 ~ M2+X "

# lavaan output
obs_output <- lavaan::sem(model = observed, sample.cov = covmatrix, sample.nobs = 200)

summary(obs_output)
#> Length  Class   Mode 
#>      1 lavaan     S4
# lavaan syntax for phantom variable model
phantom <- " M2 ~ M1 + Y1 + a*X
                Y2 ~ M1 + Y1 + b*M2 + cp*X "

Step1 <- SA_step1(
  lavoutput = obs_output,
  mod_obs = observed,
  mod_phant = phantom
)
#> Here are the phantom covariance matrix parameters (copy and paste and add values/names for step2):
#> 
#> 
#> phantom_assignment <-( "CovM1M2"= ,
#> "CovM1X"= ,
#> "CovM1Y2"= ,
#> "CovY1M1"= ,
#> "CovY1M2"= ,
#> "CovY1X"= ,
#> "CovY1Y2"= ,
#> "VarM1"= ,
#> "VarY1" = )
#> 
#>  Choose the names of the phantom covariances that you want to fix to single values and put in a vector. These will be used for the fixed_names argument in the SA_step2 function.  The phantom covariance parameters that you want to vary should be put in a list and used as the test_names argument.
#> Here are the observed covariance matrix parameters:
#> [1] "CovXM2"  "CovXY2"  "CovY2M2"
#> Choose which values you want to use for your fixed parameters and put their names in a vector (fixed_values). Make sure the order is the same for both vectors.

phantom_assignment <- list(
  "CovM1X" = 0,
  "CovY1M1" = "CovY2M2",
  "CovY1X" = 0,
  "VarM1" = 1,
  "VarY1" = 1,
  "CovM1M2" = seq(0, .6, .1),
  "CovY1Y2" = "CovM1M2",
  "CovY1M2" = seq(-.6, .6, .1),
  "CovM1Y2" = "CovY1M2"
)
Step2 <- SA_step2(
  phantom_assignment = phantom_assignment,
  step1 = Step1
)
#> [1] 1
#> [1] 2
#> [1] 3
#> [1] TRUE
#> [1] TRUE
#> [1] TRUE
#> [1] 0
#> [1] TRUE
#> [1] 0
#> [1] TRUE
#> [1] 1
#> [1] TRUE
#> [1] 1
Step3 <- SA_step3(
  step2 = Step2,
  n = 200
)

b_results <- ghost_par_ests(
  step3 = Step3,
  parameter_label = "b",
  remove_NA = TRUE
)

head(b_results)
#>            rep lhs op rhs label       est         se         z       pvalue
#> reptemp      1  Y2  ~  M2     b 1.1355601 0.09909786 11.458977 0.000000e+00
#> reptemp.7    8  Y2  ~  M2     b 0.6082658 0.10339672  5.882834 4.032996e-09
#> reptemp.8    9  Y2  ~  M2     b 1.1274865 0.08421692 13.387885 0.000000e+00
#> reptemp.14  15  Y2  ~  M2     b 0.3983970 0.10131046  3.932436 8.408924e-05
#> reptemp.15  16  Y2  ~  M2     b 0.6894810 0.09548591  7.220762 5.169198e-13
#> reptemp.16  17  Y2  ~  M2     b 1.1244882 0.07796970 14.422119 0.000000e+00
#>             ci.lower  ci.upper
#> reptemp    0.9413319 1.3297883
#> reptemp.7  0.4056119 0.8109196
#> reptemp.8  0.9624244 1.2925486
#> reptemp.14 0.1998321 0.5969618
#> reptemp.15 0.5023321 0.8766300
#> reptemp.16 0.9716704 1.2773060
```

## Help

If you encounter errors in the package, please file an issue with a
minimal reproducible example on Github or email the package maintainer.


NEWS.md
1# phantSEM (development version)

* Initial CRAN submission.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(phantSEM)

test_check("phantSEM")


tests/testthat/test-SA_step1.R
library(lavaan)

test_that("SA_step1 returns a list of length 7", {
  mod_obs <-
    " Y2 ~ X + M2
    M2 ~ X"

  mod_phant <-
    "Y2 ~ X + M1 + Y1 + M2
   M2 ~ X + M1 + Y1"

  cov <- matrix(c(1, .07, -.19, .07, 1, -.22, -.19, -.22, 1), nrow = 3, byrow = 3)
  rownames(cov) <- c("X", "M2", "Y2")
  colnames(cov) <- rownames(cov)

  lavoutput <- sem(model = mod_obs, sample.cov = cov, sample.nobs = 50)
  expect_equal(length(SA_step1(lavoutput, mod_obs, mod_phant)), 7)
})



test_that("SA_step1 has same results for sem() and lavaan(auto.var=TRUE)", {
  mod_obs <-
    " Y2 ~ X + M2
    M2 ~ X"

  mod_phant <-
    "Y2 ~ X + M1 + Y1 + M2
   M2 ~ X + M1 + Y1"

  cov <- matrix(c(1, .07, -.19, .07, 1, -.22, -.19, -.22, 1), nrow = 3, byrow = 3)
  rownames(cov) <- c("X", "M2", "Y2")
  colnames(cov) <- rownames(cov)

  lavoutputsem <- sem(model = mod_obs, sample.cov = cov, sample.nobs = 50)
  lavoutputlavaan <- lavaan(model = mod_obs, sample.cov = cov, sample.nobs = 50, auto.var = TRUE)
  expect_equal(SA_step1(lavoutputsem, mod_obs, mod_phant), SA_step1(lavoutputlavaan, mod_obs, mod_phant))
})


test_that("check that matrix template output is correct dimensions", {
  mod_obs <-
    " Y2 ~ X + M2
    M2 ~ X"

  mod_phant <-
    "Y2 ~ X + M1 + Y1 + M2
   M2 ~ X + M1 + Y1"

  cov <- matrix(c(1, .07, -.19, .07, 1, -.22, -.19, -.22, 1), nrow = 3, byrow = 3)
  rownames(cov) <- c("X", "M2", "Y2")
  colnames(cov) <- rownames(cov)

  lavoutput <- sem(model = mod_obs, sample.cov = cov, sample.nobs = 50)

  expect_equal(nrow(SA_step1(lavoutput, mod_obs, mod_phant)[[1]]), 5)
  expect_equal(nrow(SA_step1(lavoutput, mod_obs, mod_phant)[[3]]), 5)
  expect_equal(nrow(SA_step1(lavoutput, mod_obs, mod_phant)[[4]]), 5)
  expect_equal(nrow(SA_step1(lavoutput, mod_obs, mod_phant)[[7]]), 25)
})

test_that("check that symmetry is correct in output by comparing output to transpose", {
  mod_obs <-
    " Y2 ~ X + M2
    M2 ~ X"

  mod_phant <-
    "Y2 ~ X + M1 + Y1 + M2
   M2 ~ X + M1 + Y1"

  cov <- matrix(c(1, .07, -.19, .07, 1, -.22, -.19, -.22, 1), nrow = 3, byrow = 3)
  rownames(cov) <- c("X", "M2", "Y2")
  colnames(cov) <- rownames(cov)

  lavoutput <- sem(model = mod_obs, sample.cov = cov, sample.nobs = 50)
  mat <- SA_step1(lavoutput, mod_obs, mod_phant)[[4]]
  expect_equal(mat, t(mat))

  mat2 <- SA_step1(lavoutput, mod_obs, mod_phant)[[3]]
  expect_false(isTRUE(all.equal(mat2, t(mat2))))
})


test_that("check that non-nested models error out", {
  mod_obs <-
    " Y2 ~ X + M2
    M2 ~ X"

  mod_phant <-
    "Y2 ~  M1 + Y1 + M2
   M2 ~  M1 + Y1"

  cov <- matrix(c(1, .07, -.19, .07, 1, -.22, -.19, -.22, 1), nrow = 3, byrow = 3)
  rownames(cov) <- c("X", "M2", "Y2")
  colnames(cov) <- rownames(cov)

  lavoutput <- sem(model = mod_obs, sample.cov = cov, sample.nobs = 50)

  expect_error(SA_step1(lavoutput, mod_obs, mod_phant))
})


tests/testthat/test-SA_step2.R
library(lavaan)

test_that("Check number of output objects", {
  mod_obs <-
    " Y2 ~ X + M2
    M2 ~ X"

  mod_phant <-
    "Y2 ~ X + M1 + Y1 + M2
   M2 ~ X + M1 + Y1"

  cov <- matrix(c(1, .07, -.19, .07, 1, -.22, -.19, -.22, 1), nrow = 3, byrow = 3)
  rownames(cov) <- c("X", "M2", "Y2")
  colnames(cov) <- rownames(cov)
  lavoutput <- sem(model = mod_obs, sample.cov = cov, sample.nobs = 50)
  step1 <- SA_step1(lavoutput, mod_obs, mod_phant)

  phantom_assignment <- list(
    "CovM1M2" = seq(0, .4, .1),
    "CovM1X" = 0,
    "CovM1Y2" = seq(-.3, .3, .1),
    "CovY1M1" = "CovY2M2",
    "CovY1M2" = "CovM1Y2",
    "CovY1X" = 0,
    "CovY1Y2" = "CovM1M2",
    "VarM1" = 1,
    "VarY1" = 1
  )

  expect_equal(length(SA_step2(phantom_assignment, step1)), 5)
})


test_that("Check if function works when given one set of test values to be used for two phantom params", {
  mod_obs <-
    " Y2 ~ X + M2
    M2 ~ X"

  mod_phant <-
    "Y2 ~ X + M1 + Y1 + M2
   M2 ~ X + M1 + Y1"

  cov <- matrix(c(1, .07, -.19, .07, 1, -.22, -.19, -.22, 1), nrow = 3, byrow = 3)
  rownames(cov) <- c("X", "M2", "Y2")
  colnames(cov) <- rownames(cov)
  lavoutput <- sem(model = mod_obs, sample.cov = cov, sample.nobs = 50)
  step1 <- SA_step1(lavoutput, mod_obs, mod_phant)

  phantom_assignment <- list(
    "CovM1M2" = seq(0, .4, .1),
    "CovM1X" = 0,
    "CovM1Y2" = "CovM2Y2",
    "CovY1M1" = "CovY2M2",
    "CovY1M2" = "CovM2Y2",
    "CovY1X" = 0,
    "CovY1Y2" = "CovM1M2",
    "VarM1" = 1,
    "VarY1" = 1
  )

  expect_equal(length(SA_step2(phantom_assignment, step1)), 5)
})


test_that("Check that phantom variables equal to other phantom variables will error out if the phantom variable name is not written correctly", {
  mod_obs <-
    " Y2 ~ X + M2
    M2 ~ X"

  mod_phant <-
    "Y2 ~ X + M1 + Y1 + M2
   M2 ~ X + M1 + Y1"

  cov <- matrix(c(1, .07, -.19, .07, 1, -.22, -.19, -.22, 1), nrow = 3, byrow = 3)
  rownames(cov) <- c("X", "M2", "Y2")
  colnames(cov) <- rownames(cov)
  lavoutput <- sem(model = mod_obs, sample.cov = cov, sample.nobs = 50)
  step1 <- SA_step1(lavoutput, mod_obs, mod_phant)



  phantom_assignment2 <- list(
    "CovM1M2" = seq(0, .4, .1),
    "CovM1X" = 0,
    "CovM1Y2" = seq(0, .3, .1),
    "CovY1M1" = "CovM2Y2",
    "CovY1M2" = "CovY2M1",
    "CovY1X" = 0,
    "CovY1Y2" = "CovM2M1",
    "VarM1" = 1,
    "VarY1" = 1
  )


  expect_error(SA_step2(phantom_assignment2, step1))
})


test_that("Check if covariance names can be written with the variables in either order (e.g., CovY1Y2 and CovY2Y1)", {
  mod_obs <-
    " Y2 ~ X + M2
    M2 ~ X"

  mod_phant <-
    "Y2 ~ X + M1 + Y1 + M2
   M2 ~ X + M1 + Y1"

  cov <- matrix(c(1, .07, -.19, .07, 1, -.22, -.19, -.22, 1), nrow = 3, byrow = 3)
  rownames(cov) <- c("X", "M2", "Y2")
  colnames(cov) <- rownames(cov)
  lavoutput <- sem(model = mod_obs, sample.cov = cov, sample.nobs = 50)
  step1 <- SA_step1(lavoutput, mod_obs, mod_phant)

  phantom_assignment1 <- list(
    "CovM1M2" = seq(0, .4, .1),
    "CovM1X" = "CovXM2",
    "CovM1Y2" = "CovM2Y2",
    "CovY1M1" = "CovY2M2",
    "CovY1M2" = "CovM2Y2",
    "CovY1X" = "CovY2X",
    "CovY1Y2" = "CovM1M2",
    "VarM1" = 1,
    "VarY1" = 1
  )

  phantom_assignment2 <- list(
    "CovM1M2" = seq(0, .4, .1),
    "CovM1X" = "CovM2X",
    "CovM1Y2" = "CovY2M2",
    "CovY1M1" = "CovM2Y2",
    "CovY1M2" = "CovY2M2",
    "CovY1X" = "CovXY2",
    "CovY1Y2" = "CovM1M2",
    "VarM1" = 1,
    "VarY1" = 1
  )


  expect_equal(SA_step2(phantom_assignment1, step1)[[4]], SA_step2(phantom_assignment2, step1)[[4]])
  expect_equal(SA_step2(phantom_assignment1, step1), SA_step2(phantom_assignment2, step1))
})


test_that("Check that function works if numeric values are given for all phantom parameters", {
  mod_obs <-
    " Y2 ~ X + M2
    M2 ~ X"

  mod_phant <-
    "Y2 ~ X + M1 + Y1 + M2
   M2 ~ X + M1 + Y1"

  cov <- matrix(c(1, .07, -.19, .07, 1, -.22, -.19, -.22, 1), nrow = 3, byrow = 3)
  rownames(cov) <- c("X", "M2", "Y2")
  colnames(cov) <- rownames(cov)
  lavoutput <- sem(model = mod_obs, sample.cov = cov, sample.nobs = 50)
  step1 <- SA_step1(lavoutput, mod_obs, mod_phant)

  phantom_assignment <- list(
    "CovM1M2" = .4,
    "CovM1X" = 0,
    "CovM1Y2" = .5,
    "CovY1M1" = .3,
    "CovY1M2" = .1,
    "CovY1X" = 0,
    "CovY1Y2" = .4,
    "VarM1" = 1,
    "VarY1" = 1
  )

  expect_equal(length(SA_step2(phantom_assignment, step1)), 5)
})


test_that("Check that when phantom parameter is not included that the default test values are used", {
  mod_obs <-
    " Y2 ~ X + M2
    M2 ~ X"

  mod_phant <-
    "Y2 ~ X + M1 + Y1 + M2
   M2 ~ X + M1 + Y1"

  cov <- matrix(c(1, .07, -.19, .07, 1, -.22, -.19, -.22, 1), nrow = 3, byrow = 3)
  rownames(cov) <- c("X", "M2", "Y2")
  colnames(cov) <- rownames(cov)
  lavoutput <- sem(model = mod_obs, sample.cov = cov, sample.nobs = 50)
  step1 <- SA_step1(lavoutput, mod_obs, mod_phant)

  phantom_assignment <- list(
    "CovM1M2" = .4,
    "CovM1X" = 0,
    "CovM1Y2" = .5,
    "CovY1M1" = .3,
    "CovY1M2" = .1,
    "CovY1X" = 0,
    # "CovY1Y2"= .4,
    "VarM1" = 1,
    "VarY1" = 1
  )

  expect_equal(SA_step2(phantom_assignment, step1)[[3]][1], data.frame(CovY1Y2 = c(seq(-.3, .3, .1))))
})


# phantom_assignment <-list("CovM1M2"=seq(-.1,.1,.05),
#                          "CovM1X"= 0,
#                          "CovM1Y2"= .5,
#                          "CovY1M1"= .3,
#                          "CovY1M2"= .1,
#                          "CovY1X"= 0,
#                          #"CovY1Y2"= .4,
#                          "VarM1"= 1,
#                          "VarY1" = 1)




test_that("Check that function works when provided with observed variable names for all phantom variables", {
  mod_obs <-
    " Y2 ~ X + M2
    M2 ~ X"

  mod_phant <-
    "Y2 ~ X + M1 + Y1 + M2
   M2 ~ X + M1 + Y1"

  cov <- matrix(c(1, .07, -.19, .07, 1, -.22, -.19, -.22, 1), nrow = 3, byrow = 3)
  rownames(cov) <- c("X", "M2", "Y2")
  colnames(cov) <- rownames(cov)
  lavoutput <- sem(model = mod_obs, sample.cov = cov, sample.nobs = 50)
  step1 <- SA_step1(lavoutput, mod_obs, mod_phant)

  phantom_assignment <- list(
    "CovM1M2" = "CovM2Y2",
    "CovM1X" = "CovM2X",
    "CovM1Y2" = "CovM2Y2",
    "CovY1M1" = "CovM2Y2",
    "CovY1M2" = "CovM2Y2",
    "CovY1X" = "CovY2X",
    "CovY1Y2" = "CovM2Y2",
    "VarM1" = "VarM2",
    "VarY1" = "VarY2"
  )

  expect_equal(length(SA_step2(phantom_assignment, step1)), 5)
})


R/SA_step1.R
#' Sensitivity Analysis Function Step 1
#' @description
#' `SA_step1()` is used to identify the phantom variables and generate names for their covariance parameters. The output of this function will be used in SA_step2().
#'
#' @param lavoutput The lavaan output object output from lavaan functions sem() or lavaan() when fitting your observed model.
#' @param mod_obs A lavaan syntax for the observed model.
#' @param mod_phant A lavaan syntax for the phantom variable model.
#' @returns a list containing the names of all phantom covariance parameters.
#' @import lavaan
#' @importFrom lavaan sem
#' @export
#' @examples
#' # covariance matrix
#' covmatrix <- matrix(c(
#'   0.25, 0.95, 0.43,
#'   0.95, 8.87, 2.66,
#'   0.43, 2.66, 10.86
#' ), nrow = 3, byrow = TRUE)
#' colnames(covmatrix) <- c("X", "M2", "Y2")
#'
#' # lavann syntax for observed model
#' observed <- " M2 ~ X
#'              Y2 ~ M2+X "
#'
#' # lavaan output
#' obs_output <- lavaan::sem(model = observed, sample.cov = covmatrix, sample.nobs = 200)
#'
#' # lavaan syntax for phantom variable model
#' phantom <- " M2 ~ M1 + Y1 + a*X
#'                Y2 ~ M1 + Y1 + b*M2 + cp*X "
#'
#' Step1 <- SA_step1(
#'   lavoutput = obs_output,
#'   mod_obs = observed,
#'   mod_phant = phantom
#' )
#'
SA_step1 <- function(lavoutput, # lavaan object
                     mod_obs, # lavaan syntax for observe model
                     mod_phant # model with phantom variables
) {
  fit <- lavoutput
  C <- fit@SampleStats@cov[[1]]
  colnames(C) <- fit@Data@ov.names[[1]]
  var_obs <- diag(C)

  obsvar <- unique(c(lavaan::lavParseModelString(mod_obs)$lhs, lavaan::lavParseModelString(mod_obs)$rhs))
  SAvar <- unique(c(lavaan::lavParseModelString(mod_phant)$lhs, lavaan::lavParseModelString(mod_phant)$rhs))

  # check for nested
  stopifnot(
    "Your phantom model must include all variables in your observed model" =
      obsvar %in% SAvar
  )

  # find the variables that are new in the phantom model -- those are the phantom vars
  phantom_names <- setdiff(SAvar, obsvar)

  # create expanded matrix
  newmat <- matrix(NA, nrow = length(SAvar), ncol = length(SAvar))
  newmat[1:length(obsvar), 1:length(obsvar)] <- stats::cov2cor(C)

  # newnames <- c(fit@Data@ov.names[[1]],phantom_names)
  # simpnames <- gsub(mediator,"M",oldnames[[2]])

  # add var names to expanded matrix
  colnames(newmat) <- c(fit@Data@ov.names[[1]], phantom_names)
  rownames(newmat) <- colnames(newmat)
  # put variance of phantom variables along diagonal
  var_phant <- c(var_obs, c(rep(1, length(phantom_names))))

  matrix_template <- newmat

  # put names of phantom variances/covariances into expanded matrix
  parname <- c()
  for (i in 1:nrow(newmat)) {
    for (j in 1:ncol(newmat)) {
      if (is.na(newmat[i, j]) & i == j) {
        tn <- (paste0("Var", rownames(newmat)[i]))
        newmat[i, j] <- tn
        parname <- c(parname, tn)
      } else if (is.na(newmat[i, j]) & i > j) {
        tn <- (paste0("Cov", rownames(newmat)[i], colnames(newmat)[j]))
        newmat[i, j] <- tn
        parname <- c(parname, tn)
      } else if (is.na(newmat[i, j]) & i < j) {
        tn <- (paste0("Cov", rownames(newmat)[j], colnames(newmat)[i]))
        newmat[i, j] <- tn
        parname <- c(parname, tn)
      }
    }
  }

  namemat <- newmat
  cov_names <- colnames(newmat)
  namemat <- outer(cov_names, cov_names, function(x, y) {
    ifelse(x == y, paste0("Var", x),
      ifelse(x < y, paste0("Cov", x, y),
        paste0("Cov", x, y)
      )
    )
  })

  obsname <- setdiff(unique(parname), namemat[lower.tri(namemat)])


  cov_map <- data.frame(
    covname = c(namemat),
    val = c(matrix_template),
    matrix(which(namemat > 0, arr.ind = TRUE), ncol = 2)
  )

  message(paste("Here are the phantom covariance matrix parameters (copy and paste and add values/names for step2):\n
              "))
  q <- paste0('"', sort(unique(parname)), '"', collapse = "= ,\n")
  cat("phantom_assignment <-(", q, "= )")
  # print(paste0(sort(unique(parname)),collapse='","'),quote=FALSE) #old
  message("\n Choose the names of the phantom covariances that you want to fix to single values and put in a vector. These will be used for the fixed_names argument in the SA_step2 function.  The phantom covariance parameters that you want to vary should be put in a list and used as the test_names argument. ")
  message("Here are the observed covariance matrix parameters:")
  # print(sort(namemat[lower.tri(namemat)]))
  print(sort(setdiff(namemat[lower.tri(namemat)], unique(parname))))
  message("Choose which values you want to use for your fixed parameters and put their names in a vector (fixed_values). Make sure the order is the same for both vectors.")
  return(list(
    matrix_template = matrix_template,
    Phantom_covs = parname,
    named_matrix = namemat,
    obs_matrix_phant_names = newmat,
    mod_phant = mod_phant,
    variances = var_phant,
    cov_map = cov_map
  ))
}


R/SA_step3.R
#' Step 3 of sensitivity analysis function
#'
#' `SA_step3()` computes the parameter estimates in your phantom model defined in step 1 for the different values provided.
#' @param step2 The object returned from SA_step2.
#' @param n The sample size.
#' @returns A list of parameter estimates from each test covariance matrix.
#' @import lavaan
#' @importFrom lavaan sem
#' @export
#' @examples
#' #' @examples
# covariance matrix
#' covmatrix <- matrix(c(
#'   0.25, 0.95, 0.43,
#'   0.95, 8.87, 2.66,
#'   0.43, 2.66, 10.86
#' ), nrow = 3, byrow = TRUE)
#' colnames(covmatrix) <- c("X", "M2", "Y2")
#'
#' # lavann syntax for observed model
#' observed <- " M2 ~ X
#'              Y2 ~ M2+X "
#'
#' # lavaan output
#' obs_output <- lavaan::sem(model = observed, sample.cov = covmatrix, sample.nobs = 200)
#'
#' # lavaan syntax for phantom variable model
#' phantom <- " M2 ~ M1 + Y1 + a*X
#'                Y2 ~ M1 + Y1 + b*M2 + cp*X "
#'
#' Step1 <- SA_step1(
#'   lavoutput = obs_output,
#'   mod_obs = observed,
#'   mod_phant = phantom
#' )
#'
#' phantom_assignment <- list(
#'   "CovM1X" = 0,
#'   "CovY1M1" = "CovY2M2",
#'   "CovY1X" = 0,
#'   "VarM1" = 1,
#'   "VarY1" = 1,
#'   "CovM1M2" = seq(.4, .6, .1),
#'   "CovY1Y2" = "CovM1M2",
#'   "CovY1M2" = seq(.1, .3, .1),
#'   "CovM1Y2" = "CovY1M2"
#' )
#' Step2 <- SA_step2(
#'   phantom_assignment = phantom_assignment,
#'   step1 = Step1
#' )
#' Step3 <- SA_step3(
#'   step2 = Step2,
#'   n = 200
#' )
#'
SA_step3 <- function(step2, n) {
  combos <- step2[[3]]
  mod_phant <- step2[[1]]
  var_phant <- step2[[2]]
  corlist <- step2[[4]]
  covlist <- step2[[5]]

  # indices of NA in matrix

  # naind <- which((is.na(matrix)&lower.tri(matrix)),arr.ind=TRUE)



  sumlist <- list(NA)
  for (i in 1:nrow(combos)) {
    if (corlist[[i]][2] == TRUE) {
      sumlist[[i]] <- lavaan::parameterEstimates(lavaan::sem(model = mod_phant, sample.cov = covlist[[i]], sample.nobs = n))
    } else {
      sumlist[[i]] <- c("NPD")
    }
  }
  return(list(sumlist, corlist, covlist, combos))
}


R/SA_lookup.R
#' Lookup Table for Sensitivity Analysis
#'
#' `SA_lookup()` is used to look up the sensitivity analysis results for a two-wave mediation model when provided with the cross-sectional correlations.
#' @param CorXM The observed correlation between predictor X and mediator M.
#' @param CorXY The observed correlation between predictor X and outcome Y.
#' @param CorMY The observed correlation between mediator M and outcome Y.
#' @import dplyr
#' @returns Results of a sensitivity analysis with varying cross-lagged and autoregressive correlations.
#' @export
#' @examples
#' # specify correlations
#' xm <- .2
#' xy <- .3
#' my <- .4
#'
#' output <- SA_lookup(
#'   CorXM = xm,
#'   CorXY = xy,
#'   CorMY = my
#' )
SA_lookup <- function(
    CorXM,
    CorXY,
    CorMY) {
  # we have to round these -- may change in future if lookup table is larger
  xm <- round(CorXM, 1)
  xy <- round(CorXY, 1)
  my <- round(CorMY, 1)


  obs2 <- obs2
  cond_id <- obs2 %>% filter(round(obs2$corxm, 1) == xm, round(obs2$corxy, 1) == xy, round(obs2$cormy, 1) == my)

  # check to see if
  if (nrow(cond_id) == 0) {
    stop("The correlation matrix you have input results in a non-positive definite covariance matrix. This may be due to rounding. Please use `SA_step1`, `SA_step2` and `SA_step3")
  }


  filt_results <- results_smaller[which(results_smaller$cond == cond_id[, 4]), ]

  return(filt_results)
}


data-raw/Lookup_Table.R
## code to prepare `Lookup_Table` dataset goes here

usethis::use_data(Lookup_Table, overwrite = TRUE)

# create lookup table for phantom baseline longitudinal model



observedall <- expand.grid(corxm = seq(-.9, .9, .1), corxy = seq(-.9, .9, .1), cormy = seq(-.9, .9, .1))


observedall$cond <- seq(1, nrow(observedall), 1)


test <- data.frame(NA)
for (i in 1:nrow(observedall)) {
  df <- matrix(c(1, observedall[i, 1], observedall[i, 2], observedall[i, 1], 1, observedall[i, 3], observedall[i, 2], observedall[i, 3], 1), nrow = 3)
  test[i, ] <- corpcor::is.positive.definite(df)
}

observedall <- cbind(observedall, test)
observed <- observedall[which(observedall$NA. == TRUE), ]

results <- matrix(nrow = 0, ncol = 1)
for (i in 1:nrow(observed)) {
  df <- matrix(c(1, observed[i, 1], observed[i, 2], observed[i, 1], 1, observed[i, 3], observed[i, 2], observed[i, 3], 1), nrow = 3)
  temp <- cbind(cond = observed[i, 4], SA_tempbias_cov(CorStability = c(-.9, .9), CorCL = c(-.9, .9), corM1Y1 = c(-.9, .9), increments = c(.1, .1, .3), covmat = df, n = 300, extreme = TRUE))
  results <- rbind(results, temp)
}

# remove NA
results_noNA <- results %>% filter(jointsigab == "Yes" | jointsigab == "No")
setwd("G:/My Drive/ASU/Dave lab/temporal bias/Manuscript/Revision/Lookup Table")
write.csv(results, "./lookup_table_all.csv")
write_rds(results, "./lookup_table_all.rds")



write.csv(results_noNA, "./lookup_table_noNA.csv")
write_rds(results_noNA, "./lookup_table_noNA.rds")

# remove empty columns
obs2 <- observed[, 1:4]
write_rds(obs2, "./conditions.rds")


results_noNA <- readRDS("G:/My Drive/ASU/Dave lab/temporal bias/Manuscript/Revision/Lookup Table/lookup_table_noNA_saverds.rds")


#saveRDS(results_smaller,"G:/My Drive/ASU/Dave lab/temporal bias/Manuscript/Revision/Lookup Table/lookup_table_noNA_small.rds",compress="xz")
#saveRDS(results_noNA,"G:/My Drive/ASU/Dave lab/temporal bias/Manuscript/Revision/Lookup Table/lookup_table_noNA_saverds.rds",compress="xz")


results_smaller <- results_noNA[,c(1:5,9,14)]

results_smaller$cond <- as.integer(results_smaller$cond)


#results_smaller <- readRDS("G:/My Drive/ASU/Dave lab/temporal bias/Manuscript/Revision/Lookup Table/lookup_table_noNA_small.rds")

# this puts it in the package
usethis::use_data(results_smaller, obs2, internal = TRUE, overwrite=TRUE, compress = "xz")



# creates data-raw
# usethis::use_data_raw("Lookup_Table")


R/ghost_par_ests.R
#' Provide parameter estimates from sensitivity analysis function
#'
#' `ghost_par_ests()` Selects certain parameter estimates from the output of the sensitivity analysis.
#' @param step3 The object returned from SA_step3.
#' @param parameter_label The label used for the parameter in the lavaan code.
#' @param remove_NA Remove rows for combinations of phantom variable parameters that resulted in inadmissable solutions in lavaan.
#' @returns A dataframe of the parameter estimates from the lavaan model.
#' @importFrom stats na.omit
#' @export
#'
#' @examples
#' # example code
#'
#' covmatrix <- matrix(c(
#'   0.25, 0.95, 0.43,
#'   0.95, 8.87, 2.66,
#'   0.43, 2.66, 10.86
#' ), nrow = 3, byrow = TRUE)
#' colnames(covmatrix) <- c("X", "M2", "Y2")
#'
#' # lavann syntax for observed model
#' observed <- " M2 ~ X
#'              Y2 ~ M2+X "
#'
#' # lavaan output
#' obs_output <- lavaan::sem(model = observed, sample.cov = covmatrix, sample.nobs = 200)
#'
#' # lavaan syntax for phantom variable model
#' phantom <- " M2 ~ M1 + Y1 + a*X
#'                Y2 ~ M1 + Y1 + b*M2 + cp*X "
#'
#' Step1 <- SA_step1(
#'   lavoutput = obs_output,
#'   mod_obs = observed,
#'   mod_phant = phantom
#' )
#'
#' phantom_assignment <- list(
#'   "CovM1X" = 0,
#'   "CovY1M1" = "CovY2M2",
#'   "CovY1X" = 0,
#'   "VarM1" = 1,
#'   "VarY1" = 1,
#'   "CovM1M2" = seq(.4, .6, .1),
#'   "CovY1Y2" = "CovM1M2",
#'   "CovY1M2" = seq(.2, .4, .1),
#'   "CovM1Y2" = "CovY1M2"
#' )
#' Step2 <- SA_step2(
#'   phantom_assignment = phantom_assignment,
#'   step1 = Step1
#' )
#' Step3 <- SA_step3(
#'   step2 = Step2,
#'   n = 200
#' )
#'
#' b_results <- ghost_par_ests(
#'   step3 = Step3,
#'   parameter_label = "b",
#'   remove_NA = TRUE
#' )
ghost_par_ests <- function(step3,
                           parameter_label,
                           remove_NA = FALSE) {
  resultlist <- step3[[1]]
  combos <- step3[[4]]
  parmatrix <- matrix(NA, nrow = 0, ncol = 10)
  comboindex <- matrix(NA, nrow = 0, ncol = 1)
  colnames(parmatrix) <- c(
    "lhs",
    "op",
    "rhs",
    "label",
    "est",
    "se",
    "z",
    "pvalue",
    "ci.lower",
    "ci.upper"
  )

  for (i in 1:nrow(combos)) {
    reptemp <- i
    if (length(resultlist[[i]]) > 1) {
      pars <- resultlist[[i]]
      partemp <- pars[which(pars$"label" == parameter_label), ]
    } else {
      partemp <- rep(NA, 10)
    }
    parmatrix <- rbind(parmatrix, partemp)
    comboindex <- rbind(comboindex, reptemp)
  }

  out <- cbind(comboindex, parmatrix)
  colnames(out) <-
    c(
      "rep",
      "lhs",
      "op",
      "rhs",
      "label",
      "est",
      "se",
      "z",
      "pvalue",
      "ci.lower",
      "ci.upper"
    )
  # out <- cbind(combos, out)

  if (remove_NA == TRUE) {
    out <- stats::na.omit(out)
  }

  return(out)
}
